src/liballoc/collections/btree/map.rs

```diff
@@ -1606,6 +1641,76 @@ impl<K, V> Clone for Values<'_, K, V> {
     }
 }
 
+/// An iterator produced by calling `drain_filter` on BTreeMap.
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+pub struct DrainFilter<'a, K, V, F>
+where
+    K: 'a + Ord,
+    V: 'a,
+    F: 'a + FnMut(&K, &mut V) -> bool,
+{
+    pred: F,
+    retained: usize,
+    length: &'a mut usize,
+    cur_leaf_edge: Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge>,
+}
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, K, V, F> Drop for DrainFilter<'a, K, V, F>
+where
+    K: 'a + Ord,
+    V: 'a,
+    F: 'a + FnMut(&K, &mut V) -> bool,
+{
+    fn drop(&mut self) {
+        self.for_each(drop);
+    }
+}
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, K, V, F> fmt::Debug for DrainFilter<'a, K, V, F>
+where
+    K: 'a + fmt::Debug + Ord,
+    V: 'a + fmt::Debug,
+    F: 'a + FnMut(&K, &mut V) -> bool,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("DrainFilter").field(&self.length).finish()
+    }
+}
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, K, V, F> Iterator for DrainFilter<'a, K, V, F>
+where
+    K: 'a + Ord,
+    V: 'a,
+    F: 'a + FnMut(&K, &mut V) -> bool,
+{
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<(K, V)> {
+        while self.retained < *self.length {
+            let (next_leaf_edge, kv, k, v) =
+                unsafe { ptr::read(&self.cur_leaf_edge).next_unchecked_kv() };
```

Comment:
Why doesn't this just use `next_unchecked()` to update the edge directly? It can still be fixed during removal if needed.

Reply:
We need that kv handle, in order to remove_kv on it, so we need an extra function that not only duplicates handles but exposes both, so it seems most elegant to just return both.

Reply:
Or in other words, no it cannot be fixed back to its old value right now, because we don't know what that was after removing something on the right. It's still the left leaf edge handle, but we gave up our kv handle too.

Actually, if we did have this, isn't `next_unchecked()` the wrong abstraction for this and for the iterators as well? First, we want the `right_kv_treewise` (draft name), descending or ascending if the local `right_kv` hits the end, without duplicating any node reference. Later, iterators want to take a further half step from the KV handle for the `right_leaf_edge`, without duplicating any node reference. In between, iterators duplicate the KV handle themselves to extract references when appropriate. The DrainFilter iterator would reuse `right_kv_treewise` and the predicate decides whether to move on to the `right_leaf_edge`.

Reply:
I mean that last bit as in [this PR candidate](https://github.com/rust-lang/rust/compare/master...ssomers:btree_navigation_revisited). For this PR, that means is that the right_kv_treewise_unchecked function (or a wrapper) would be reused here, and the scary duplication of marker::Mut handles would happen inside DrainFilter.

Reply:
Benchmark performance was up 20% to 200% over many benches, so I wondered if making the functions safe (which used to cost 5%- 10%) could be acceptable this way. Amazingly, it sprinkles around another 20% to 100%.

So for this PR, that means right_kv_treewise is already there to be reused.